### DISCUSSION TITLE
Mongoose Query resolves even if it wasn't awaited

### DISCUSSION THREAD

**Discussion by samislam** [Automattic/mongoose#12172] [Q&A] [answered]
I have the following Model:

```js
const userSchema = new mongoose.Schema({
  // long long schema... etc
  age: Number,
})
const UserModel = mongoose.model('User', userSchema, 'users')
```
I have the following function:
```js
async function x(y) {
  // @param y: function | asyncfunction
  const YreturnValue = await y()
  YreturnValue.sort('age')
  return YreturnValue // returns a Mongoose Query
}
```
and I have the following route:
```js
router.route('/').get(async (req, res, next) => {
  const query = await x(() => UserModel.find({}))
  const docs = await query 
  res.status(200).json({ data: docs })
})
```

When I request the `GET /`, I get this error:
[![enter image description here][1]][1]

I am *awaiting* the function **x** because it's an async function (i.e, it returns a Promise), and the resolved value of the function **x** is **expected** to be a **mongoose Query object**.

- the function **x** receives one argument, a function, or an async function, in both cases, your argument must return a mongoose Query object.

What's the problem? 

I have asked this question on StackOverflow, but I didn't get any fulfilling response.
https://stackoverflow.com/posts/73166696/edit

  [1]: https://i.stack.imgur.com/t1l2U.png

**Comment by josegl** [comment]
I think that the problem comes because you are awaiting the `y` function inside the `x` function. That will return the query results, but I guess that what you want actually is to sort the query, and for such case you should act like this:

```javascript
// @param y: function | asyncfunction
function x(y) {
  return new Promise((resolve, reject) => {
    y.sort('age').exec((err, results) => {
      if(err){
        reject(err);
      }else{
        resolve(results);
      }
    });
  });  
}
```
In your version:
```javascript
async function x(y) {
  // @param y: function | asyncfunction
  const YreturnValue = await y()
  YreturnValue.sort('age')
  return YreturnValue // returns a Mongoose Query
}
```
When you `await` the `y` function it resolves to the query results, so `YreturnValue` is not a Query as you think, it is an array, and when you want to sort, what you are sorting is an array, so it fails because the sort function expects a function or undefined.

You maybe could try this version that I'm not sure if it would work:
```javascript
// @param y: function | asyncfunction
async function x(y) {
  sortedQuery = y.sort('age');
  const YreturnValue = await sortedQuery();
  return YreturnValue // returns an Array
}
```
If what you want is to return a Query instance with the sort configuration, then you could try:
```javascript
// @param y: function | asyncfunction
async function x(y) {
  sortedQuery = y.sort('age');
  return sortedQuery;
}
```
And then here:
```javascript
router.route('/').get(async (req, res, next) => {
  const query = await x(() => UserModel.find({}))
  const docs = await query 
  res.status(200).json({ data: docs })
})
```
`query` will be an actual Mongoose Query with the `sort` configuration applied. However as I said above, I'm not sure if you can `await` a sorted query and get the results, or if you have to use the first implementation that I suggested.

**Comment by vkarpov15** [comment, marked as answer]
If you `return` something that has a `then()` function from an async function, JavaScript will automatically call `then()`. There's no way to return a thenable from an async function.

You would have to wrap your query in an object, like:

```javascript
async function x(y) {
  // @param y: function | asyncfunction
  const YreturnValue = await y()
  YreturnValue.sort('age')
  return { query: YreturnValue } // returns a Mongoose Query
}
```
